src/transform/src/nonnull_requirements.rs

```diff
@@ -195,14 +195,52 @@ impl NonNullRequirements {
                 expected_group_size: _,
             } => {
                 let mut new_columns = HashSet::new();
-                for column in columns {
-                    // No obvious requirements on aggregate columns.
-                    // A "non-empty" requirement, I guess?
-                    if column < group_key.len() {
-                        group_key[column].non_null_requirements(&mut new_columns);
-                    }
-                    if column == group_key.len() && aggregates.len() == 1 {
-                        aggregates[0].expr.non_null_requirements(&mut new_columns);
+                let (group_key_columns, aggs_columns): (Vec<usize>, Vec<usize>) =
+                    columns.iter().partition(|c| **c < group_key.len());
+                for column in group_key_columns {
+                    group_key[column].non_null_requirements(&mut new_columns);
+                }
+
+                if !aggs_columns.is_empty() {
+                    // We can convert a non-null requirement on an aggregate function into
+                    // a non-null requirement on its parameter only if we are sure that
+                    // won't affect the result of any other aggregation.
```

Comment:
Can you say more about this? I would have thought we want the union of the constraints. The intent of the analysis is that we can safely drop any row that exhibits a null value in any of the listed columns, in which case preserving the values of other columns seems irrelevant.

Reply:
We can't union constraints as the following example shows:

```
materialize=> select t1.f1, sum(t1.f2), max(t2.f2) from t1 left join t2 on t1.f2 = t2.f1 group by t1.f1 having max(t2.f2) >= 0;
 f1 | sum | max 
----+-------+-----
  1 |     2 |   2
(1 row)
materialize=> select * from t1 left join t2 on t1.f2 = t2.f1;
 f1 | f2 | f1 | f2 
----+----+----+----
  1 |  1 |    |   
  1 |  1 |    |   
  1 |  0 |  0 |  2
(3 rows)
materialize=> 
```

`max(t2.f2)` is required to be non-null, but we can't require `t2.f2` to be non-null, because that would result in `sum(t1.f2)` being 0, rather than 2 in that example, due to the null rows being removed.

We can push down the non-null requirement iff there is no other aggregation which result might be affected.

Instead of computing the union, we intersect the sets of columns that are required for the input of all the aggregations to be non-null. We can require the input columns of the non-null aggregations to be non-null iff no other aggregation may have a non-null input from a row with null values in all the columns in the non-null aggregation, ie. all the columns in the non-null aggregation must be present in that intersection of columns.

Reply:
I'll have to re-read that last paragraph a few times, but to double check our understanding: the intent of the transformation is that it is pushing down the information "this row will be dropped if it contains a `null` in any of these columns". Under that interpretation, I don't yet understand why we intersect the requirements.

There is some moment at which we *introduce* new requirements, for example with an empty non-null requirements, but a reduce who plans to ignore some rows with null columns. I agree that in that case we want to intersect the constraints, but it isn't what the text above in the comment says ("We can convert a non-null requirement on an aggregate function into a non-null requirement on its parameter ... ").

It looks like the code previously only triggered in the single-aggregation case, where intersection is a no-op; was it incorrect, or just missed some optimization?

Reply:
For fixing the wrong results issue this ticket is about, we just need the changes in `relation/mod.rs` and use that new `AggregateExpr::non_null_requirements` method in the old code for the single-aggregation case.

However, with this patch I'm also fixing the missing optimization opportunities, where all the aggregates are on a common set of columns. 

Sorry for the unclear comments. I'll try explaining them better :)

Reply:
> There is some moment at which we introduce new requirements, for example with an empty non-null requirements, but a reduce who plans to ignore some rows with null columns. I agree that in that case we want to intersect the constraints...

That is not what this code is doing because that wouldn't be correct in SQL, and that's why the comment doesn't say that, I think.

Reply:
The intent of the transformation is to push down the information "rows with null values in any of these columns won't change the result of the reduction and hence, can be discarded at optimization time"

Reply:
I think I understand what is going on now (or at least, retract my concern). What I meant wrt the above comment is that e.g. `count` is able to propose that some columns be added to the non-null requirements, even if the count output column is not itself required to be non-null, which is the sort of place these constraints come from in the first place. Most other aggregates *propagate* constraints. And once we have all of these in one place, we do want to intersect them, you are right.

Reply:
oh, you're right.

```
# outer join to inner join conversion allowed
query T multiline
EXPLAIN select t1.f1, count(t2.f2) from t1 LEFT JOIN t2 ON t1.f1 = t2.f1 group by t1.f1;
----
%0 =
| Get materialize.public.t1 (u1)
| Filter !(isnull(#0))
| ArrangeBy (#0)

%1 =
| Get materialize.public.t2 (u3)
| Filter !(isnull(#0))

%2 = Let l0 =
| Join %0 %1 (= #0 #2)
| | implementation = Differential %1 %0.(#0)
| | demand = (#0, #3)

%3 =
| Get materialize.public.t1 (u1)

%4 =
| Get %2 (l0)
| Distinct group=(#0)
| ArrangeBy (#0)

%5 =
| Join %3 %4 (= #0 #2)
| | implementation = Differential %3 %4.(#0)
| | demand = (#0)
| Negate
| Project (#0, #1)

%6 =
| Get materialize.public.t1 (u1)

%7 =
| Union %5 %6
| Map null, null

%8 =
| Union %7 %2
| Reduce group=(#0)
| | agg count(#3)

EOF

# outer join to inner join conversion allowed
query T multiline
EXPLAIN select t1.f1, count(t2.f2), max(t2.f2) from t1 LEFT JOIN t2 ON t1.f1 = t2.f1 group by t1.f1;
----
%0 =
| Get materialize.public.t1 (u1)
| Filter !(isnull(#0))
| ArrangeBy (#0)

%1 =
| Get materialize.public.t2 (u3)
| Filter !(isnull(#0))

%2 = Let l0 =
| Join %0 %1 (= #0 #2)
| | implementation = Differential %1 %0.(#0)
| | demand = (#0, #3)

%3 =
| Get materialize.public.t1 (u1)

%4 =
| Get %2 (l0)
| Distinct group=(#0)
| ArrangeBy (#0)

%5 =
| Join %3 %4 (= #0 #2)
| | implementation = Differential %3 %4.(#0)
| | demand = (#0)
| Negate
| Project (#0, #1)

%6 =
| Get materialize.public.t1 (u1)

%7 =
| Union %5 %6
| Map null, null

%8 =
| Union %7 %2
| Reduce group=(#0)
| | agg count(#3)
| | agg max(#3)

EOF

# outer join to inner join conversion allowed
query T multiline
EXPLAIN select t1.f1, count(t2.f2), max(t2.f2) from t1 LEFT JOIN t2 ON t1.f1 = t2.f1 group by t1.f1 having max(t2.f2) > 0;
----
%0 =
| Get materialize.public.t1 (u1)
| Filter !(isnull(#0))
| ArrangeBy (#0)

%1 =
| Get materialize.public.t2 (u3)
| Filter !(isnull(#0))

%2 = Let l0 =
| Join %0 %1 (= #0 #2)
| | implementation = Differential %1 %0.(#0)
| | demand = (#0, #3)

%3 =
| Get materialize.public.t1 (u1)

%4 =
| Get %2 (l0)
| Distinct group=(#0)
| ArrangeBy (#0)

%5 =
| Join %3 %4 (= #0 #2)
| | implementation = Differential %3 %4.(#0)
| | demand = (#0)
| Negate
| Project (#0, #1)

%6 =
| Get materialize.public.t1 (u1)

%7 =
| Union %5 %6
| Map null, null

%8 =
| Union %7 %2
| Reduce group=(#0)
| | agg count(#3)
| | agg max(#3)
| Filter (#2 > 0)

EOF
```

In those three cases we should be removing the outer join, but we aren't.

Reply:
I don't want to admit to knowing if the above does / doesn't check out, but another case I could imagine is 
```
SELECT
    max(col1),
    max(col2),
    max(col3),
FROM ..
```
whose columns don't intersect, but which can still remove outer joins. Though, not easily in the language of `nonnull_requirements`, which is communicating an "and" requirements on columns (any null can cause a drop) vs an "or" (if all columns are null, can drop).

It seems like maybe we should eventually roll this in to a predicate pushdown, which I think you've suggested before.

Reply:
btw, don't let all of these comments get in the way of 1. fixing bugs and 2. making strict improvements that may not capture everything.